### PR TITLE
Persistent log

### PR DESCRIPTION
**Pull request by jmftrindade**

### CODE REVIEW

**Comment by jonhoo** [comment]
This merge seems wrong in a couple of places. For example, `targets/netsoup.rs` shouldn't exist..?

**Comment by jmftrindade** [comment]
Yup, it looks like these 3 extra files are artifacts of previous merges (it's likely they stayed around because the original changes on master moved them elsewhere, and they weren't being used anymore?):

- benchmarks/vote/targets/netsoup.rs
- src/flow/data.rs
- src/flow/sql_to_flow.rs

Will remove these from the branch, make sure that all tests still pass, and update this request.

**Comment by jmftrindade** [comment]
Alright, pull request updated.

File: Cargo.toml [hunk 1]
```diff
@@ -9,9 +9,9 @@ b_memcached = ["memcached-rs"]
 b_postgresql = ["postgres"]
 b_mysql = ["mysql"]
 b_mssql = ["futures", "futures-state-stream", "tiberius", "tokio-core"]
-b_netsoup = ["futures", "tokio-core", "tarpc", "tarpc-plugins", "serde", "serde_derive"]
+b_netsoup = ["futures", "tokio-core", "tarpc", "tarpc-plugins"]
 b_hybrid = ["mysql", "memcached-rs"]
-default = ["web", "b_netsoup", "b_memcached", "b_mysql", "b_mssql", "b_hybrid"]
+default = ["web", "b_netsoup", "b_memcached", "b_mysql", "b_mssql", "b_hybrid", "serde", "serde_derive"]
```
Comment (jonhoo) [inline Cargo.toml]: At this point you can probably just make `serde` and `serde_derive` be non-optional dependencies, and then not list them under `default`.

File: src/ops/base.rs [hunk 2]
```diff
@@ -1,39 +1,221 @@
...
+    /// Synchronous writes: forces every record to be written to disk before it is emitted further
+    /// into the data-flow graph. Strong guarantees (writes are never lost), but high performance
+    /// penalty.
+    SyncImmediately,
+}
+
 impl Base {
     /// Create a base node operator.
-    pub fn new(primary_key: Vec<usize>) -> Self {
+    pub fn new(primary_key: Vec<usize>, durability: BaseDurabilityLevel) -> Self {
```
Comment (jonhoo) [inline src/ops/base.rs]: Can we instead add another method, `new_durable`, that takes two arguments, leaving `new` to just take one, and give no durability? Would also reduce the diff since all the tests don't have to change :p

File: src/flow/core/processing.rs [hunk 3]
```diff
@@ -84,7 +84,7 @@ pub trait Ingredient
                 data: prelude::Records,
                 domain: &prelude::DomainNodes,
                 states: &prelude::StateMap)
-                -> prelude::Records;
+                -> Option<prelude::Records>;
```
Comment (jonhoo) [inline src/flow/core/processing.rs]: Agh, that's going to be annoying to incorporate into `partial-materialization`, but I guess it's necessary here.

File: src/flow/domain/single.rs [hunk 4]
```diff
@@ -161,7 +161,7 @@ impl NodeDescriptor {
             }
             flow::node::Type::Internal(ref mut i) => {
                 let from = m.link().src;
-                m.map_data(|data| i.on_input(from, data, nodes, state));
+                m.map_data(|data| i.on_input(from, data, nodes, state).unwrap());
```
Comment (jonhoo) [inline src/flow/domain/single.rs]: Why is this `unwrap()` okay? Can't a `Base` be processed?

File: src/flow/mod.rs [hunk 5]
```diff
@@ -463,7 +463,7 @@ impl<'a> Migration<'a> {
                 self.mainline.ingredients.add_edge(*parent.as_global(), ni, false);
             }
         }
-        // and tell the caller its id
+
```
Comment (jonhoo) [inline src/flow/mod.rs]: Why?

File: src/flow/payload.rs [hunk 6]
```diff
@@ -74,6 +74,9 @@ pub enum Packet {
         parents: Vec<LocalNodeIndex>,
     },
 
+    /// Instruct base nodes to flush any buffered writes.
```
Comment (jonhoo) [inline src/flow/payload.rs]: Can we cut this entirely for now?

File: src/lib.rs [hunk 7]
```diff
@@ -175,7 +175,8 @@
 //! # let mut g = Blender::new();
 //! # let article = {
 //! # let mut mig = g.start_migration();
-//! # let article = mig.add_ingredient("article", &["id", "title"], Base::default());
+//! # let article = mig.add_ingredient("article", &["id", "title"],
+//! #                                  Base::default());
```
Comment (jonhoo) [inline src/lib.rs]: Why?

File: src/ops/base.rs [hunk 8]
```diff
@@ -1,39 +1,221 @@
...
 /// type corresponding to the node's type.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Base {
+    buffered_writes: Option<Records>,
+    durability: BaseDurabilityLevel,
+    durable_log: Option<BufWriter<File, WhenFull>>,
+    durable_log_path: Option<PathBuf>,
+    global_address: Option<NodeAddress>,
+    last_flushed_at: Option<Instant>,
```
Comment (jonhoo) [inline src/ops/base.rs]: A lot of `Option` here. Could we instead have a `struct BaseBufferState`, and then have `Base` just have one field of type `Option<BaseBufferState>`?

File: src/ops/base.rs [hunk 9]
```diff
@@ -1,39 +1,221 @@
...
+            BaseDurabilityLevel::None => panic!("tried to persist non-durable base node!"),
+            BaseDurabilityLevel::Buffered |
+            BaseDurabilityLevel::SyncImmediately => {
+                self.ensure_log_writer();
+                serde_json::to_writer(&mut self.durable_log.as_mut().unwrap(), &records)
+                    .unwrap();
+                // XXX(malte): we must deconstruct the BufWriter in order to get at the contained
+                // File (on which we can invoke sync_data(), only to then reassemble it
+                // immediately. I suspect this will work best if we flush after accumulating
+                // batches of writes.
```
Comment (jonhoo) [inline src/ops/base.rs]: Does dropping a `BufWriter` not flush the underlying stream? If so, `drop(xxx.take())` should be sufficient.

File: src/ops/base.rs [hunk 10]
```diff
@@ -1,39 +1,221 @@
...
+                Err(reason) => {
+                    panic!("Unable to open durable log file {}, reason: {}",
+                           path.display(), reason)
+                }
+                Ok(file) => file,
+            };
+
+            match self.durability {
+                BaseDurabilityLevel::None => {
+                    self.durable_log = None;
```
Comment (jonhoo) [inline src/ops/base.rs]: In this case it seems like we also shouldn't even create the log file, no?

File: src/ops/base.rs [hunk 11]
```diff
@@ -1,39 +1,221 @@
...
+                }
+            }
+        }
+    }
+
+    /// XXX: This should only be used by tests.  We don't hide it behind cfg test, however, since it
+    /// needs to be available for integration tests, which get compiled against the regular build.
+    pub fn delete_durable_log(&mut self) {
+        // Cleanup any durable log files.
+        if self.durable_log_path.is_some() {
```
Comment (jonhoo) [inline src/ops/base.rs]: `if let Some(ref path) = self.durable_log_path` avoids the `unwrap` below.

File: src/ops/base.rs [hunk 12]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
                 rs: Records,
                 _: &DomainNodes,
                 state: &StateMap)
-                -> Records {
-        rs.into_iter()
+                -> Option<Records> {
+        // Write incoming records to log before processing them if we are a durable node.
+        let records_to_return;
+        let mut copy_rs = rs.clone();
```
Comment (jonhoo) [inline src/ops/base.rs]: Why is this `clone` necessary? It seems pretty costly, especially because it's not just executed for durable nodes.

File: src/ops/base.rs [hunk 13]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+                //
+                // 1. Enough time has passed since the last time we flushed.
+                // 2. Our buffer of write records reaches capacity.
+                let num_buffered_writes = self.buffered_writes.as_ref().unwrap().len();
+                let has_reached_capacity = num_buffered_writes + rs.len() >= BUFFERED_WRITES_CAPACITY;
+                let elapsed = self.last_flushed_at.unwrap().elapsed();
+                let has_reached_time_limit = elapsed >= Duration::from_millis(BUFFERED_WRITES_FLUSH_INTERVAL_MS);
+
+                if has_reached_capacity || has_reached_time_limit {
+                    self.buffered_writes.as_mut().unwrap().append(copy_rs.as_mut());
```
Comment (jonhoo) [inline src/ops/base.rs]: Can't this just take `rs`?

File: src/ops/base.rs [hunk 14]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+
+                    let copy_buffered_writes = self.buffered_writes.as_mut().unwrap().clone();
```
Comment (jonhoo) [inline src/ops/base.rs]: Can't this just `drain(..)`?

File: src/ops/base.rs [hunk 15]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+                    self.flush();
+
+                    // This returns everything that was buffered, plus the newly inserted records.
+                    records_to_return = Some(copy_buffered_writes);
+                } else {
+                    // Otherwise, buffer the records and don't send them downstream.
+                    self.buffered_writes.as_mut().unwrap().append(copy_rs.as_mut());
```
Comment (jonhoo) [inline src/ops/base.rs]: This can also just consume `rs`

File: src/ops/base.rs [hunk 16]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+                    return Some(Records::default())
```
Comment (jonhoo) [inline src/ops/base.rs]: Shouldn't this return `None`?

File: src/ops/base.rs [hunk 17]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+                }
+            },
+            BaseDurabilityLevel::SyncImmediately => {
+                self.persist_to_log(&rs);
+                records_to_return = Some(rs.clone());
```
Comment (jonhoo) [inline src/ops/base.rs]: Why does this need to clone?

File: src/ops/base.rs [hunk 18]
```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
...
+            },
+            BaseDurabilityLevel::None => {
+                records_to_return = Some(rs.clone());
```
Comment (jonhoo) [inline src/ops/base.rs]: ditto

File: src/ops/mod.rs [hunk 19]
```diff
@@ -190,7 +190,9 @@ pub mod test {
                 .clear();
         }
 
-        pub fn one<U: Into<Records>>(&mut self, src: NodeAddress, u: U, remember: bool) -> Records {
+        pub fn one<U: Into<Records>>(&mut self,
```
Comment (jonhoo) [inline src/ops/mod.rs]: Unrelated?

File: src/ops/mod.rs [hunk 20]
```diff
@@ -204,7 +206,8 @@ pub mod test {
                 .as_local()]
                     .borrow_mut()
                     .inner
-                    .on_input(src, u.into(), &self.nodes, &self.states);
+                    .on_input(src, u.into(), &self.nodes, &self.states)
```
Comment (jonhoo) [inline src/ops/mod.rs]: Unrelated?

File: src/ops/mod.rs [hunk 21]
```diff
@@ -190,7 +190,9 @@ pub mod test {
                 .clear();
         }
 
-        pub fn one<U: Into<Records>>(&mut self, src: NodeAddress, u: U, remember: bool) -> Records {
+        pub fn one<U: Into<Records>>(&mut self,
+                                     src: NodeAddress,
+                                     u: U, remember: bool) -> Records {
```
Comment (jonhoo) [inline src/ops/mod.rs]: Just formatting?

File: tests/lib.rs [hunk 22]
```diff
@@ -71,6 +73,178 @@ fn it_works() {
...
 #[test]
+fn it_propagates_writes_w_durability_sync_immediately() {
+    use distributary::{Base, BaseDurabilityLevel};
+
+    // set up graph
+    let mut g = distributary::Blender::new();
+    let (a, b, cq) = {
+        let mut mig = g.start_migration();
+        let a = mig.add_ingredient("a", &["a", "b"],
+                                   Base::new(vec![0], BaseDurabilityLevel::SyncImmediately).delete_log_on_drop());
```
Comment (jonhoo) [inline tests/lib.rs]: To make this look a bit neater, you may want to assign `BaseDurabilityLevel::SyncImmediately` to a variable with a short name instead :)

File: tests/lib.rs [hunk 23]
```diff
@@ -136,16 +310,16 @@ fn shared_interdomain_ancestor() {
     // send a value on a
     muta.put(vec![id.clone(), 2.into()]);
     assert_eq!(bq.recv_timeout(time::Duration::from_millis(100)),
-               Ok(vec![vec![id.clone(), 2.into()].into()]));
+    Ok(vec![vec![id.clone(), 2.into()].into()]));
```
Comment (jonhoo) [inline tests/lib.rs]: just formatting?

File: tests/lib.rs [hunk 24]
```diff
@@ -208,12 +382,15 @@ fn it_works_w_mat() {
...
+
     // set up graph
     let mut g = distributary::Blender::new();
     let (a, b, cq) = {
         let mut mig = g.start_migration();
-        let a = mig.add_ingredient("a", &["x", "y"], distributary::Base::new(vec![1]));
-        let b = mig.add_ingredient("b", &["_", "x", "y"], distributary::Base::new(vec![2]));
+        let a = mig.add_ingredient("a", &["x", "y"], Base::new(vec![1], BaseDurabilityLevel::None));
+        let b = mig.add_ingredient("b", &["_", "x", "y"],
+                                   Base::new(vec![2], BaseDurabilityLevel::None));
```
Comment (jonhoo) [inline tests/lib.rs]: With the `::new` change, this diff will disappear.

File: tests/lib.rs [hunk 25]
```diff
@@ -241,7 +418,7 @@ fn it_works_deletion() {
     use distributary::StreamUpdate::*;
     muta.delete(vec![2.into()]);
     assert_eq!(cq.recv(),
-               Ok(vec![DeleteRow(Arc::new(vec![1.into(), 2.into()]))]));
+    Ok(vec![DeleteRow(Arc::new(vec![1.into(), 2.into()]))]));
```
Comment (jonhoo) [inline tests/lib.rs]: Just formatting?

File: tests/lib.rs [hunk 26]
```diff
@@ -623,7 +800,7 @@ fn transactional_migration() {
 
     // check that c got them
     assert_eq!(cq(&3.into()).unwrap().0,
-               vec![vec![3.into(), 5.into()], vec![3.into(), 6.into()]]);
+    vec![vec![3.into(), 5.into()], vec![3.into(), 6.into()]]);
```
Comment (jonhoo) [inline tests/lib.rs]: Just formatting?

File: tests/lib.rs [hunk 27]
```diff
@@ -994,7 +1171,7 @@ fn live_writes() {
         let mut mig = g.start_migration();
 
         // add vote base table
-        vote = mig.add_ingredient("vote", &["user", "id"], distributary::Base::default());
+        vote = mig.add_ingredient("vote", &["user", "id"], Base::default());
```
Comment (jonhoo) [inline tests/lib.rs]: I'd prefer it if you removed this diff, just to make the PR smaller.

**Comment by jonhoo** [comment]
Okay, @jmftrindade, I think that's all I wanted to say :p Overall looks pretty good — this is mostly just nit.

File: Cargo.toml (hunk 1 above)
Comment (jmftrindade) [inline Cargo.toml]: Done

File: src/ops/base.rs (hunk 2 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done

File: src/flow/mod.rs (hunk 5 above)
Comment (jmftrindade) [inline src/flow/mod.rs]: Reverted

File: src/lib.rs (hunk 7 above)
Comment (jmftrindade) [inline src/lib.rs]: Reverted

File: tests/lib.rs (hunk 25 above)
Comment (jmftrindade) [inline tests/lib.rs]: Reverted

File: tests/lib.rs (hunk 26 above)
Comment (jmftrindade) [inline tests/lib.rs]: Reverted

File: tests/lib.rs (hunk 27 above)
Comment (jmftrindade) [inline tests/lib.rs]: Done

File: tests/lib.rs (hunk 24 above)
Comment (jmftrindade) [inline tests/lib.rs]: Done

File: src/ops/mod.rs (hunk 20 above)
Comment (jmftrindade) [inline src/ops/mod.rs]: This one removes the semicolon from this line, and adds an unwrap() call right after.

File: src/ops/mod.rs (hunk 21 above)
Comment (jmftrindade) [inline src/ops/mod.rs]: Reverted

File: src/ops/base.rs [hunk 28]
```diff
@@ -1,39 +1,234 @@
...
+    /// penalty.
+    SyncImmediately,
+}
+
 impl Base {
-    /// Create a base node operator.
+    /// Create a non-durable base node operator.
     pub fn new(primary_key: Vec<usize>) -> Self {
         Base {
+            buffered_writes: Some(Records::default()),
```
Comment (jonhoo) [inline src/ops/base.rs]: Personal preference, but I'd prefer if this and `new_durable` both internally called a single method that takes an `Option<BaseDurabilityLevel>`. That way we don't have to repeat the initialization of all the fields.

File: src/ops/base.rs [hunk 29]
```diff
@@ -1,39 +1,234 @@
...
+            last_flushed_at: Some(Instant::now()),
+            primary_key: Some(primary_key),
+            should_delete_log_on_drop: false,
+            unique_id: ProcessUniqueId::new(),
+            us: None,
+        }
+    }
+
+    /// Create durable base node operator.
+    pub fn new_durable(primary_key: Vec<usize>, durability: BaseDurabilityLevel) -> Self {
```
Comment (jonhoo) [inline src/ops/base.rs]: Couldn't we do the `ensure_log_writer` stuff here? And that way remove the call to it entirely from `on_input`?

File: src/ops/base.rs (hunk 28 above)
Comment (jonhoo) [inline src/ops/base.rs]: Or alternatively, they could both construct using `Base::default()`, and then modify the appropriate fields.

File: src/flow/domain/single.rs (hunk 4 above)
Comment (jmftrindade) [inline src/flow/domain/single.rs]: Base::on_input is returning Option<Records> now, hence the unwrap().  Should we have a match here instead?

Comment (jonhoo) [inline src/flow/domain/single.rs]: Well, `unwrap` indicates that you believe `on_input` can *never* return `None` in this context. Why is that?

File: src/ops/base.rs (hunk 8 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Do you think it will still be better than what we have now, given that not all durable fields are always consistently None / Some?  For example, what we currently have in Base::new_durable for the durability state fields is:
...
            buffered_writes: Some(Records::default()),
            durability: Some(durability),
            durable_log: None,
            durable_log_path: None,
            last_flushed_at: Some(Instant::now()),
...

Which means we can't make the fields required on BaseDurabilityState, i.e.,:

struct BaseDurabilityState {
    buffered_writes: Records,
    durability: BaseDurabilityLevel,
    durable_log: BufWriter<File, WhenFull>,
    durable_log_path: PathBuf,
    last_flushed_at: Instant,
}

So we'd have 2 layers of optional fields.  Wouldn't that be weird?

Comment (jonhoo) [inline src/ops/base.rs]: Hmm.. Isn't `buffered_writes` and `durability` only interesting if `BaseDurability != None`, and therefore it's fine for them to be inside `BaseDurabilityState` without an `Option`? Same for `last_flushed_at`, which is only used if durability `!= None`?

File: src/ops/base.rs (hunk 29 above)
Comment (jmftrindade) [inline src/ops/base.rs]: The durable log path filename relies on self.us local id, which is not available at construction time: it is only provided by the time we get to Base::on_commit.

Comment (jonhoo) [inline src/ops/base.rs]: Ah, that's a shame. But fair enough, I guess.

File: src/ops/base.rs (hunk 8 above)
Comment (jmftrindade) [inline src/ops/base.rs]: I see.  I think `durable_log_path` would still need to be optional, since at construction time we don't have the Base node's "us" id (it's only provided at `::on_commit`).  Because `durable_log_path` is not known at creation time, `durable_log` would also need to be optional -- otherwise if we instead relied on some default value for path, we'd unnecessarily create a durable log file that's not gonna be used.

Comment (jonhoo) [inline src/ops/base.rs]: Oh, so this stems from the same issue below of not knowing the node's ID until later. That's unfortunate. I think I'd still prefer `Option`s inside of a `BaseDurabilityState`, but you're right that it's less elegant. I wonder if there's a way we could name `Base` nodes when they're created instead of relying on `self.us`. That would mean we could instantiate all of this at creation time, get rid of the `Option`s, and also get rid of the call to `ensure_log`...

File: src/ops/base.rs (hunk 10 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Yeah, this code looks weird.  So the caller (`::persist_to_log`) was already taking care of calling `::ensure_log_writer` only for the appropriate durability levels.  Moved `::ensure_log_writer`'s match one level up, and made the `None` option unreachable.

File: src/ops/base.rs (hunk 8 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Yes, naming `Base` nodes at creation time would be ideal, and it would also make it easier for a recovery manager to restore `Base` nodes -- since now an external entity can keep track of the global_id -> Base node mapping, it knows where to fetch the corresponding durable log for a given node.

How does global id generation currently work right now?  Could we expose that as a thread-safe singleton somewhere, so that every time we call `Base::new` it gets assigned an unique pseudo random number (LCG, xorshift, etc)?

Comment (jonhoo) [inline src/ops/base.rs]: Currently each node is assigned a `NodeIndex` when it is inserted into the `petgraph::Graph`, so it's monotonically increasing at `add_ingredient` time. An alternative would be to add a required `name` argument to `Base::new`, which we require to be unique...

File: src/flow/domain/single.rs (hunk 4 above)
Comment (jmftrindade) [inline src/flow/domain/single.rs]: Because `::on_input` never actually returned None, only Some(empty_records_vector) when buffered writes weren't propagated downstream.  Updated.

File: src/flow/payload.rs (hunk 6 above)
Comment (jmftrindade) [inline src/flow/payload.rs]: Done.

File: src/ops/base.rs (hunk 8 above)
Comment (jmftrindade) [inline src/ops/base.rs]: But the `NodeIndex` is "local", right?  Adding a `name` argument to `Base::new` that we "assume" to be unique  is a good short-term solution vs perhaps a long-term one being some thread-safe unique PRNG.

The latter makes sure Soup users are not responsible for providing unique names, and that base nodes don't have to check whether their provided `name` is unique.

Alright, will keep this as is.  In a separate pull request -- where we name `Base` nodes -- I can revisit the `Option`s "overload" here :)

File: src/ops/base.rs (hunk 28 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done

File: src/ops/base.rs (hunk 11 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done

File: src/ops/base.rs (hunk 13 above)
Comment (jmftrindade) [inline src/ops/base.rs]: `Vec::append` expects a mutable reference, and what we get in the arg list is a vanilla `rs: Records`. If I instead pass rs as a `&mut` (`self.buffered_writes.as_mut().unwrap().append(&mut rs)`), I get a  "cannot borrow mutably" compilation error.

Once I changed `::on_input` to expect a mutable rs instead, then it all worked without the cloning.

File: src/ops/base.rs (hunk 14 above)
Comment (jmftrindade) [inline src/ops/base.rs]: `drain(..)` would consume the underlying vector, right?  We still need the original buffered records to remain there for the `flush` call on the next line.

File: src/ops/base.rs (hunk 15 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done

File: src/ops/base.rs (hunk 16 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Yeah, that makes a lot more sense than Some(empty Vec).  Otherwise, `::on_input` could just return a regular records Vec.

So I updated callers to explicitly handle `Base::on_input` returning None in the case of buffered writes that are not propagated downstream.

Oddly enough I now get a panic from `Project::resolve_col` via `it_fails_to_resolve_literal`, but this seems unrelated since there's no explicit handling of `Base::on_input` results there?  Why would the check for `self.emit.is_some() && col >= self.emit.as_ref().unwrap().len()` be true if `Base::on_input` can now return `None`?

File: src/ops/base.rs (hunk 17 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Removed

File: src/ops/base.rs (hunk 18 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done

File: src/ops/mod.rs (hunk 20 above)
Comment (jmftrindade) [inline src/ops/mod.rs]: Updated this to explicitly handle `Base::on_input` returning `None` -- instead of implicitly expecting it to return Some(empty vector).

File: tests/lib.rs (hunk 22 above)
Comment (jmftrindade) [inline tests/lib.rs]: Done here and for `BaseDurabilityLevel::Buffered` test below.

File: tests/lib.rs (hunk 23 above)
Comment (jmftrindade) [inline tests/lib.rs]: Reverted

File: src/ops/base.rs (hunk 9 above)
Comment (jmftrindade) [inline src/ops/base.rs]: It does, but we still need the original `file` to create a the new `BufWriter`.  So we can instead drop `file.sync_data()` -- since we need the `file` anyway on the next line.

File: src/ops/base.rs (hunk 12 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Yeah, it isn't.  Wrote it like that because (1) I clearly haven't gotten a good grasp of Rust's borrowing semantics yet, and (2) other permutations of this that I tried all showed cryptic compilation error messages :)  Removed excessive cloning here.

File: src/flow/domain/single.rs (hunk 4 above)
Comment (jonhoo) [inline src/flow/domain/single.rs]: Hmm, if never returns `None`, why does it need to return an `Option`?

File: src/ops/base.rs (hunk 8 above)
Comment (jonhoo) [inline src/ops/base.rs]: `NodeIndex` is global for a given execution, `NodeAddress` can be (thread)local or global for the execution. Yeah, I like the idea of moving to `name` — will probably make some of this a bit nicer to work with :) In a separate PR seems fine though.

File: src/ops/base.rs (hunk 12 above)
Comment (jonhoo) [inline src/ops/base.rs]: Hehe, we've all been there :)

File: src/ops/base.rs (hunk 14 above)
Comment (jonhoo) [inline src/ops/base.rs]: Ah, no, `drain(..)` crucially does *not* consume the vector, though it does *empty* it. I guess the way you might get away without this clone is to have `self.flush` return the vec it flushed, so that we can then return it again. Sorry to keep poking at these `clones`, but I think it might end up having a decent impact given how all writes go through `Base` nodes.

File: src/ops/base.rs (hunk 16 above)
Comment (jonhoo) [inline src/ops/base.rs]: Oh, that's weird -- I don't see how that test should be at all related to changes to `Base` :/

**Comment by jonhoo** [comment]
One question: if we *always* end up treating `None` from `on_input` as `Records::default()`, why don't we just keep it a `Records` like it used to be, and return that directly? Will make it easier to merge now following the merge of `partial-materialization` as well..

**Comment by jonhoo** [comment]
Also, don't worry too much about the conflicts with `master`. I can do that merge when it comes to that :)

**Comment by jonhoo** [comment]
Oh, nice, that significantly reduced the number of conflicts too! :D

File: src/flow/domain/single.rs (hunk 4 above)
Comment (jmftrindade) [inline src/flow/domain/single.rs]: Resolved

File: src/ops/base.rs (hunk 16 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Ok, so that was actually a false alarm.  That test is not failing, as it expects a panic -- should have been obvious from the `it_fails_to` prefix hehe.  It's just that I run my tests with RUST_BACKTRACE enabled, and since I was getting a backtrace for it, I incorrectly assumed it was a test failure :)

File: src/ops/base.rs (hunk 11 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Actually done now.

File: src/ops/base.rs (hunk 14 above)
Comment (jmftrindade) [inline src/ops/base.rs]: Done! 💃

**Comment by jonhoo** [comment]
@jmftrindade I'll take a stab at merging `master` into this at some point tonight, so if you suddenly get a push error, that'll be why. Hopefully it shouldn't be too tricky to rebase any changes you make in the interim on top of that merge.

**Comment by jonhoo** [comment]
Travis failures are due to https://github.com/rust-lang-nursery/rustup.rs/issues/1062

**Comment by jonhoo** [comment]
:tada:
